packages/SwingSet/test/virtualObjects/test-representatives.js

```diff
@@ -151,3 +157,168 @@ test('virtual object representatives', async t => {
   ]);
   t.deepEqual(c.kpResolution(rz2), capdata('"overflow"'));
 });
+
+test('exercise cache', async t => {
+  const config = {
+    bootstrap: 'bootstrap',
+    vats: {
+      bootstrap: {
+        sourceSpec: path.resolve(__dirname, 'vat-representative-bootstrap.js'),
+        creationOptions: {
+          virtualObjectCacheSize: 3,
+        },
+      },
+    },
+  };
+
+  const log = [];
+
+  const storage = initSwingStore().storage;
+  function vsKey(key) {
+    return key.match(/^\w+\.vs\./);
+  }
+  const loggingStorage = {
+    has: key => storage.has(key),
+    getKeys: (start, end) => storage.getKeys(start, end),
+    get(key) {
+      const result = storage.get(key);
+      if (vsKey(key)) {
+        log.push(['get', key, result]);
+      }
+      return result;
+    },
+    set(key, value) {
+      if (vsKey(key)) {
+        log.push(['set', key, value]);
+      }
+      storage.set(key, value);
+    },
+    delete(key) {
+      if (vsKey(key)) {
+        log.push(['delete', key]);
+      }
+      storage.delete(key);
+    },
+  };
+
+  const bootstrapResult = await initializeSwingset(config, [], loggingStorage);
+  const c = await makeSwingsetController(loggingStorage, {});
+
+  const nextLog = makeNextLog(c);
+
+  await c.run();
+  t.deepEqual(c.kpResolution(bootstrapResult), capargs('bootstrap done'));
+
+  async function doSimple(method, what, ...args) {
+    let sendArgs;
+    if (what) {
+      const whatArg = {
+        '@qclass': 'slot',
+        iface: 'Alleged: thing',
+        index: 0,
+      };
+      sendArgs = capargs([whatArg, ...args], [what]);
+    } else {
+      sendArgs = capargs(args);
+    }
+    const r = c.queueToVatExport('bootstrap', 'o+0', method, sendArgs);
+    await c.run();
+    t.is(c.kpStatus(r), 'fulfilled');
+    t.deepEqual(nextLog(), []);
+    return r;
+  }
+
+  async function make(name, holdIt, expect) {
+    const r = await doSimple('makeThing', null, name, holdIt);
+    const result = c.kpResolution(r);
+    t.deepEqual(result, slot0('thing', expect));
+    return result.slots[0];
+  }
+  async function read(what, expect) {
+    const r = await doSimple('readThing', what);
+    t.deepEqual(c.kpResolution(r), capargs(expect));
+  }
+  async function readHeld(expect) {
+    const r = await doSimple('readHeldThing', null);
+    t.deepEqual(c.kpResolution(r), capargs(expect));
+  }
+  async function write(what, newName) {
+    await doSimple('writeThing', what, newName);
+  }
+  async function writeHeld(newName) {
+    await doSimple('writeHeldThing', null, newName);
+  }
+  async function forgetHeld() {
+    await doSimple('forgetHeldThing', null);
+  }
+  async function hold(what) {
+    await doSimple('holdThing', what);
+  }
+  function thingID(num) {
+    return `v1.vs.o+1/${num}`;
+  }
+  function thingVal(name) {
+    return JSON.stringify({
+      name: capdata(JSON.stringify(name)),
+    });
+  }
+
+  // init cache - []
```

Comment:
nice test structure!

We need to test that reanimating an object without reading or writing its properties will refrain from doing a syscall. Is that what line 318 does when it does `await hold('ko32');` and then asserts the log is empty? If so, awesome, let's add a comment explaining that's what we're checking there.

This structure might be easier to follow if you define a batch of `const T1 = 'ko25';` first, with a comment that these are the expected allocations, and that they might change if the kernel allocation behavior changes (e.g. if we switched to a different starting offset). Making the command lines say `await make('thing1', true, T1);` would make it easier to 1: tell what the expected correlation is, and 2: change those allocations to match any future kernel changes. I've updated calls like this in the past and it's a real drag, a lot of error-prone cutting-and-pasting. Having a single place where the "we just happen to know it's supposed to be `ko25`" value is declared would make that easier.

Reply:
Reanimation happens when an object is deserialized, which happens in the test that's there now.  The test includes cases where the reference is deserialized but then never accessed, so I think the test case you are asking for is covered.

I will do the thing with the symbolic names for the koids.